### ISSUE TITLE
The accuracy problem after streaming implementation

### ISSUE THREAD

**Issue by qingqingxu2020** [Rikorose/DeepFilterNet#106]
Hello, thanks for your open source DeepFilterNet work. After trying it out, I found the effect and the amount of computation to be excellent.

After studying the network carefully, I confirmed that it meets the requirements of streaming speech processing. Therefore, after calculating the padding size, I changed the forward inference part of the model into streaming realization (for loop). 
`
class Encoder(nn.Module):
    def __init__(self):
        super().__init__()
        p = ModelParams()
        layer_width = p.conv_ch
        wf = p.conv_width_f
        assert p.nb_erb % 4 == 0, "erb_bins should be divisible by 4"

        k = p.conv_k_enc
        kwargs = {"batch_norm": True, "depthwise": p.conv_depthwise}
        k0 = 1 if k == 1 and p.conv_lookahead == 0 else max(2, k)
        cl = 1 if p.conv_lookahead > 0 else 0
        self.erb_conv0 = convkxf(1, layer_width, k=k0, fstride=1, lookahead=cl, **kwargs)
        cl = 1 if p.conv_lookahead > 1 else 0
        self.erb_conv1 = convkxf(
            layer_width * wf**0, layer_width * wf**1, k=k, lookahead=cl, **kwargs
        )
        cl = 1 if p.conv_lookahead > 2 else 0
        self.erb_conv2 = convkxf(
            layer_width * wf**1, layer_width * wf**2, k=k, lookahead=cl, **kwargs
        )
        self.erb_conv3 = convkxf(
            layer_width * wf**2, layer_width * wf**2, k=k, fstride=1, **kwargs
        )
        self.df_conv0 = convkxf(
            2, layer_width, fstride=1, k=k0, lookahead=p.conv_lookahead, **kwargs
        )
        self.df_conv1 = convkxf(layer_width, layer_width * wf**1, k=k, **kwargs)
        self.erb_bins = p.nb_erb
        self.emb_dim = layer_width * p.nb_erb // 4 * wf**2
        self.df_fc_emb = GroupedLinear(
            layer_width * p.nb_df // 2, self.emb_dim, groups=p.lin_groups
        )
        self.emb_out_dim = p.emb_hidden_dim
        self.emb_n_layers = p.emb_num_layers
        self.gru_groups = p.gru_groups
        self.emb_gru = GroupedGRU(
            self.emb_dim,
            self.emb_out_dim,
            num_layers=p.emb_num_layers,
            batch_first=False,
            groups=p.gru_groups,
            shuffle=p.group_shuffle,
            add_outputs=True,
        )
        self.lsnr_fc = nn.Sequential(nn.Linear(self.emb_out_dim, 1), nn.Sigmoid())
        self.lsnr_scale = p.lsnr_max - p.lsnr_min
        self.lsnr_offset = p.lsnr_min

        self.streaming_state = {
            'e1': None,
            'e2': None,
            'c0': None,
        }

    def forward(
        self, feat_erb: Tensor, feat_spec: Tensor
    ) -> Tuple[Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor]:
        # Encodes erb; erb should be in dB scale + normalized; Fe are number of erb bands.
        #streaming implementation
        B, C, T, Ferb = feat_erb.shape
        erb_padding_right = torch.zeros((B, C, 2, Ferb), dtype=feat_erb.dtype, device=feat_erb.device)
        feat_erb = torch.cat([feat_erb, erb_padding_right], dim=-2)

        B, _, T, Fspec = feat_spec.shape
        spec_padding_right = torch.zeros((B, 2, 2, Fspec), dtype=feat_spec.dtype, device=feat_spec.device)
        feat_spec = torch.cat([feat_spec, spec_padding_right], dim=-2)
        e0_s, e1_s, e2_s, e3_s = None, None, None, None
        emb_s, c0_s, c1_s, lsnr_s = None, None, None, None

        self.streaming_state['e1'] = torch.zeros((B, 64, 1, Ferb // 2), dtype=feat_erb.dtype, device=feat_erb.device)
        self.streaming_state['e2'] = torch.zeros((B, 64, 1, Ferb // 4), dtype=feat_erb.dtype, device=feat_erb.device)
        self.streaming_state['c0'] = torch.zeros((B, 64, 1, Fspec), dtype=feat_spec.dtype, device=feat_spec.device)
      
        for t in range(T):  
            sub_e0 = self.erb_conv0[1:](feat_erb[:,:,t:t+3,:])  # [B, C, 2, F]
            sub_e1 = self.erb_conv1[1:](sub_e0)  # [B, C*2, 1, F/2]
            sub_e2 = self.erb_conv2[1:](torch.cat([self.streaming_state['e1'],sub_e1], dim=-2))  # [B, C*4, 1, F/4]
            sub_e3 = self.erb_conv3[1:](torch.cat([self.streaming_state['e2'],sub_e2], dim=-2))  # [B, C*4, 1, F/4]
            self.streaming_state['e1'] = sub_e1
            self.streaming_state['e2'] = sub_e2
            sub_c0 = self.df_conv0[1:](feat_spec[:,:,t+1:t+3,:])# [B, C, 1, Fc]
            sub_c1 = self.df_conv1[1:](torch.cat([self.streaming_state['c0'],sub_c0], dim=-2)) # [B, C*2, 1, Fc]  
            self.streaming_state['c0'] = sub_c0
            sub_cemb = sub_c1.permute(2, 0, 1, 3).reshape(1, B, -1)  # [1, B, C * Fc/4]
            sub_cemb = self.df_fc_emb(sub_cemb)  # [1, B, C * F/4]
            sub_emb = sub_e3.permute(2, 0, 1, 3).reshape(1, B, -1)  # [1, B, C * F/4]
            sub_emb = sub_emb + sub_cemb
            sub_emb, _ = self.emb_gru(sub_emb)
            sub_emb = sub_emb.transpose(0, 1) # [B, 1, C * F/4]
            sub_lsnr = self.lsnr_fc(sub_emb) * self.lsnr_scale + self.lsnr_offset

            if t == 0:
                e0_s, e1_s, e2_s, e3_s = sub_e0[:, :, [0], :], sub_e1, sub_e2, sub_e3
                c0_s, c1_s, emb_s, lsnr_s = sub_c0, sub_c1, sub_emb, sub_lsnr
            else:
                e0_s = torch.cat((e0_s, sub_e0[:, :, [0], :]), dim=-2)
                e1_s = torch.cat((e1_s, sub_e1), dim=-2)
                e2_s = torch.cat((e2_s, sub_e2), dim=-2)
                e3_s = torch.cat((e3_s, sub_e3), dim=-2)
                c0_s = torch.cat((c0_s, sub_c0), dim=-2)
                c1_s = torch.cat((c1_s, sub_c1), dim=-2)
                emb_s = torch.cat((emb_s, sub_emb), dim=-2)
                lsnr_s = torch.cat((lsnr_s, sub_lsnr), dim=-2)
   
        return e0_s, e1_s, e2_s, e3_s, emb_s, c0_s, lsnr_s

class ErbDecoder(nn.Module):
    def __init__(self):
        super().__init__()
        p = ModelParams()
        layer_width = p.conv_ch
        wf = p.conv_width_f
        assert p.nb_erb % 8 == 0, "erb_bins should be divisible by 8"

        self.emb_width = layer_width * wf**2
        self.emb_dim = self.emb_width * (p.nb_erb // 4)
        self.fc_emb = nn.Sequential(
            GroupedLinear(
                p.emb_hidden_dim, self.emb_dim, groups=p.lin_groups, shuffle=p.group_shuffle
            ),
            nn.ReLU(inplace=True),
        )
        k = p.conv_k_dec
        kwargs = {"k": k, "batch_norm": True, "depthwise": p.conv_depthwise}
        tkwargs = {
            "k": k,
            "batch_norm": True,
            "depthwise": p.convt_depthwise,
            "mode": p.conv_dec_mode,
        }
        pkwargs = {"k": 1, "f": 1, "batch_norm": True}
        # convt: TransposedConvolution, convp: Pathway (encoder to decoder) convolutions
        self.conv3p = convkxf(layer_width * wf**2, self.emb_width, **pkwargs)
        self.convt3 = convkxf(self.emb_width, layer_width * wf**2, fstride=1, **kwargs)
        self.conv2p = convkxf(layer_width * wf**2, layer_width * wf**2, **pkwargs)
        self.convt2 = convkxf(layer_width * wf**2, layer_width * wf**1, **tkwargs)
        self.conv1p = convkxf(layer_width * wf**1, layer_width * wf**1, **pkwargs)
        self.convt1 = convkxf(layer_width * wf**1, layer_width * wf**0, **tkwargs)
        self.conv0p = convkxf(layer_width, layer_width, **pkwargs)
        self.conv0_out = convkxf(layer_width, 1, fstride=1, k=k, act=nn.Sigmoid())

        self.streaming_state = {
            'convt3in': None,
            'convt2in': None,
            'convt1in': None,
            'conv0in': None
        }

    def forward(self, emb, e3, e2, e1, e0) -> Tensor:
        # Estimates erb mask
        #streaming implementation
        B, C, T, F8 = e3.shape
        data_type, device = e3.dtype, e3.device
        self.streaming_state['convt3in'] = torch.zeros((B, C, 1, F8), dtype=data_type, device=device)
        self.streaming_state['convt2in'] = torch.zeros((B, C, 1, F8), dtype=data_type, device=device)
        self.streaming_state['convt1in'] = torch.zeros((B, C, 1, F8*2), dtype=data_type, device=device)
        self.streaming_state['conv0in'] = torch.zeros((B, C, 1, F8*4), dtype=data_type, device=device)
        m = None
        for t in range(T):
            sub_emb = self.fc_emb(emb[:, [t], :])
            sub_emb = sub_emb.view(B, 1, -1, F8).transpose(1, 2)  # [B, C*8, T, F/8]
            convt3_in_cur = self.conv3p(e3[:, :, [t], :]) + sub_emb
            convt3_in = torch.cat([self.streaming_state['convt3in'], convt3_in_cur], dim=-2)
            self.streaming_state['convt3in'] = convt3_in_cur
            sub_e3 = self.convt3[1:](convt3_in)  # [B, C*4, T, F/4]
            convt2_in_cur = self.conv2p(e2[:, :, [t], :]) + sub_e3
            convt2_in = torch.cat([self.streaming_state['convt2in'], convt2_in_cur], dim=-2)
            self.streaming_state['convt2in'] = convt2_in_cur
            sub_e2 = self.convt2[1:](convt2_in)  # [B, C*2, T, F/2]
            convt1_in_cur = self.conv1p(e1[:, :, [t], :]) + sub_e2
            convt1_in = torch.cat([self.streaming_state['convt1in'], convt1_in_cur], dim=-2)
            self.streaming_state['convt1in'] = convt1_in_cur
            sub_e1 = self.convt1[1:](convt1_in)  # [B, C, T, F]
            conv0_in_cur = self.conv0p(e0[:, :, [t], :]) + sub_e1
            conv0_in = torch.cat([self.streaming_state['conv0in'], conv0_in_cur], dim=-2)
            self.streaming_state['conv0in'] = conv0_in_cur
            sub_m = self.conv0_out[1:](conv0_in)  # [B, 1, T, F]
            if t == 0:
                m = sub_m
            else:
                m = torch.cat((m, sub_m), dim=-2)
        return m

class DfNet(nn.Module):
    def __init__(
        self,
        erb_inv_fb: Tensor,
        run_df: bool = True,
        train_mask: bool = True,
    ):
        super().__init__()
        p = ModelParams()
        layer_width = p.conv_ch
        assert p.nb_erb % 8 == 0, "erb_bins should be divisible by 8"
        self.freq_bins = p.fft_size // 2 + 1
        self.emb_dim = layer_width * p.nb_erb
        self.erb_bins = p.nb_erb
        self.enc = Encoder()
        self.erb_dec = ErbDecoder()
        self.mask = Mask(erb_inv_fb, post_filter=p.mask_pf)

        self.df_order = p.df_order
        self.df_bins = p.nb_df
        self.df_lookahead = p.df_lookahead
        self.df_dec = DfDecoder()
        self.df_op = torch.jit.script(
            DfOp(
                p.nb_df,
                p.df_order,
                p.df_lookahead,
                freq_bins=self.freq_bins,
                method=p.dfop_method,
            )
        )

        self.run_df = run_df
        if not run_df:
            from loguru import logger
            logger.warning("Runing without DF")
        self.train_mask = train_mask

    def forward(
        self,
        spec: Tensor,
        feat_erb: Tensor,
        feat_spec: Tensor,  # Not used, take spec modified by mask instead
        atten_lim: Optional[Tensor] = None,
    ) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
        feat_spec = feat_spec.transpose(1, 4).squeeze(4)  # re/im into channel axis
        e0, e1, e2, e3, emb, c0, lsnr = self.enc(feat_erb, feat_spec)
        m = self.erb_dec(emb, e3, e2, e1, e0)
        spec = self.mask(spec, m, atten_lim)
        self.run_df = False
        if self.run_df:
            df_coefs, df_alpha = self.df_dec(emb, c0)
            spec = self.df_op(spec, df_coefs, df_alpha)
        else:
            df_alpha = torch.zeros(spec.shape[0], spec.shape[2], 1, device=spec.device)
        return spec, m, lsnr, df_alpha   
`

I only use Encoder and ErbDecoder modules. However, my result was not so good. Later, I found that the difference is due to nn.GRU inference. There will be different results between batched and one by one inference of nn.GRU because of NUMERICAL ACCURACY. https://pytorch.org/docs/stable/notes/numerical_accuracy.html

**Comment by qingqingxu2020** [comment]
already solved

**Comment by stonelazy** [comment]
@qingqingxu2020 If I can ask, how is this solved ?

**Comment by isgursoy** [comment]
https://github.com/Rikorose/DeepFilterNet/issues/105#issuecomment-1157714362
